plugins/primus_propagate_taint/primus_propagate_taint_main.ml

```diff
@@ -41,49 +41,58 @@ module Intro(Machine : Primus.Machine.S) = struct
 
   open Machine.Syntax
 
-
   let kind t =
-    Kind.create (sprintf "user-%s" (Tid.name (Term.tid t)))
+    Kind.create (sprintf "user-%s" (Tid.name t))
 
   let connect tid taint =
     Machine.Local.update mapper ~f:(fun m -> {
           m with
           tids = Map.add m.tids ~key:taint ~data:tid
         })
 
+  let translation = [
+    Legacy_taint.reg, `Reg, Taint.Rel.direct;
+    Legacy_taint.ptr, `Ptr, Taint.Rel.indirect;
+  ]
+
 
   let gentaint t =
     kind t >>= fun k ->
     Object.create k >>= fun taint ->
-    connect (Term.tid t) taint >>| fun () ->
+    connect t taint >>| fun () ->
     Taint.Object.Set.singleton taint
 
   let introduces def =
-    Term.has_attr def Legacy_taint.reg ||
-    Term.has_attr def Legacy_taint.ptr
+    List.find_map translation ~f:(fun (k,s,r) ->
+        match Term.get_attr def k with
+        | None -> None
+        | Some t -> Some (t,s,r))
 
-  let taint_ptr taints ptr sz =
-    Eval.exp ptr >>= fun ptr ->
-    Machine.Seq.iter (Seq.range 0 (Size.in_bytes sz)) ~f:(fun off ->
-        Value.nsucc ptr off >>= fun ptr ->
-        Tracker.attach ptr Taint.Rel.indirect taints)
+  let taint_ptr taints ptr =
+    Tracker.attach ptr Taint.Rel.indirect taints
 
-  let taint_var taints var =
+  let taint_var k taints var =
     Env.get var >>= fun v ->
-    Tracker.attach v Taint.Rel.direct taints
+    Tracker.attach v k taints
 
   let intro def =
-    if introduces def then
-      gentaint def >>= fun t ->
-      taint_var t (Def.lhs def) >>= fun () ->
+    match introduces def with
+    | None -> Machine.return ()
+    | Some (taint,kind,rel) ->
+      gentaint taint >>= fun t ->
+      taint_var rel t (Def.lhs def) >>= fun () ->
```

Comment:
here we taint `lhs` part, that is not completely correct in case of `store`: we taint a whole memory in this case.
So we should ignore explicitly memory taints.